content/blog/adopting-existing-cloud-resources-into-pulumi/index.md

```diff
@@ -0,0 +1,242 @@
+---
+title: "Adopting Existing Cloud Resources into Pulumi"
+date: "2019-08-15"
+meta_desc: "Adopt existing cloud infrastructure from Terraform, CloudFormation, ARM or YAML into Pulumi using `import`."
+meta_image: "console.png"
+authors: ["luke-hoban"]
+tags: ["Infrastructure"]
+---
+
+Most cloud infrastrucutre projects involve working with existing cloud resources - either building on top of existing resources or adopting existing resources under management of a new and more robust infrastructure provisioning solution.
+
+Whether you are adopting resources that were deployed manually using your cloud provider's console or CLI - or migrating existing infrastructure from tools like Terraform or CloudFormation - Pulumi makes it easy to adopt and manage your existing resources.
+
+<!--more-->
+
+When working with existing resources, there are typically two scenarios:
+1. You need to reference existing resources to use as inputs to new resources in Pulumi
+1. You need to adopt existing resources under management so they can be updated/managed in Pulumi
+
+We'll review referencing existing resources, and then dive deeper into how you can adopt existing resources with Pulumi.
+
+## Referencing Existing Resources
+For referencing existing resoruces, Pulumi offers several tools.  
+
+* The `.get` methods available on every resource let you [get all the details for a resource](https://www.pulumi.com/docs/reference/pkg/nodejs/pulumi/aws/ec2/#Vpc-get) from the cloud provider based just on it's `id`.
+* The `StackReference` resource lets you reference outputs of another stack for use as inputs to another stack, which is very useful for [organizing projects and stack](https://www.pulumi.com/docs/reference/organizing-stacks-projects/). 
+* The [`terraform.state.RemoteStateReference()`](https://www.pulumi.com/blog/using-terraform-remote-state-with-pulumi/), [`aws.cloudformation.getStack()`](https://www.pulumi.com/docs/reference/pkg/nodejs/pulumi/aws/cloudformation/#getStack) and [`azure.core.TemplateDeployment.get()`](https://www.pulumi.com/docs/reference/pkg/nodejs/pulumi/azure/core/#TemplateDeployment-get) let you reference outputs from existing Terraform, CloudFormation and ARM deployments respectively.
+
+Together, these make it easy to reference existing infrastructure regardless of how it was provisioned.
+
+## Adopting Existing Resources
+
+For adopting resources, Pulumi offers the [`import`](https://www.pulumi.com/docs/reference/pkg/nodejs/pulumi/azure/core/#TemplateDeployment-get) resource option to request that a resource defined in your Pulumi program adopt an existing resource in the cloud provider instead of creating a new one.  In keeping with it's focus on infrastructure as *code*, Pulumi lets you specify this `import` behaviour inside the Pulumi code for you infrastructure deployment, instead of outside it in some manual workflow.  In it's simplest form looks like this:
```

Comment:
```suggestion
For adopting existing resources, Pulumi offers the [`import`](https://www.pulumi.com/docs/reference/pkg/nodejs/pulumi/azure/core/#TemplateDeployment-get) resource option to request that a resource defined in your Pulumi program adopt an existing resource in the cloud provider instead of creating a new one.  In keeping with it's focus on infrastructure as *code*, Pulumi lets you specify this `import` behavior inside the Pulumi code for you infrastructure deployment, instead of outside it in some manual workflow.  In it's simplest form looks like this:
```